internal/bgp/bgp.go

```diff
@@ -665,3 +679,33 @@ func buildTCPMD5Sig(addr net.IP, key string) tcpmd5sig {
 
 	return t
 }
+
+// localAddressExists returns true if the provided address exists locally on
+// any of the host's network interfaces. The function doesn't return errors and
+// instead returns false if an error has occurred while looking for the
+// address. This function is safe to call if addr is nil or malformed.
```

Comment:
If by "using" you mean try to call `unix.Bind()` with the provided source address, to do so I would have to significantly alter the `dialMD5()` function, perform a potentially-failing syscall, "swallow" the error and then retry with `[::]`. Note that `dialMD5()` currently isn't covered by unit tests.

It seems better to me to check if the address is valid before we go through the process of constructing the socket and calling `unix.Bind()` for the following reasons:

- It seems safer to me.
- It's already implemented and works well.

Do you see a reason to switch to your bind-based approach?

Reply:
@rata and I have discussed this offline and decided that in case the user provides a non-existent IP address we shouldn't fall back to `::` but rather let `unix.Bind()` fail with an error.

Reply:
Thanks!